### ISSUE TITLE
Improved string interpolation

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#1478]
String interpolations are great, but they could be even greater by taking inspiration from collection literals.

We currently allow any *expression* inside an interpolation block (`${...}`). If we instead treated interpolations like *elements* of a collection of `String`s, then we could allow `for`, `if` and spread elements too. 

Example (one with everything):
```dart
var jsonText = """
{
${for (var key in map.keys) if (!key.startsWith('_')) ...['  ', key, ":", map[key], '\n']}
}
""";
```

One use-case I often run into is optionally adding some text to the string:
```dart
  "example${name != null ? "($name)" : ""}: something"
```
Here the `: ""` is annoying and superfluous. With `if`-interpolation we can write it as:
```dart
  "example${if (name != null) "($name)"}: something"
```

**Comment by tatumizer** [comment]
In code generation (which includes json) it would be more convenient to introduce a special kind of literal, where the compiler removes all whitespace automatically, but allows `\n', `\s`, and `\ns` (`\4s` inserts 4 spaces)
```dart
var jsonText = code"""{
for (var key in map.keys) 
   if (!key.startsWith('_')) 
      \2s $key: ${map[key]} \n
}
""";
```

**Comment by lrhn** [comment]
Or just do:
```dart
var first = true;
String sep() {
  if (!first) return ",";
  first = false;
  return "";
}
var jsonText = """{${
  for (var key in map.keys) 
     if (!key.startsWith('_')) 
        "${sep()}\n  $key: ${map[key]},\n"
}}""";
```
(Adding the correct commas is still the biggest issue. Probably want to create a helper class for it.)

**Comment by tatumizer** [comment]
The advantage of any of these forms over currently used idiom `[...].join()` is not obvious. What is it? Performance?

**Comment by lrhn** [comment]
Mostly performance, yes. 
No need to create the list, just to throw it away again, and no need to create the intermediate string from `.join()`, it can just all be added to the underlying `StringBuffer` (or whatever) and joined only once at the end.

The list is not part of the operation, it's just an accident of implementation because it's the easiest way to introduce separators automatically. That means that it is mostly noise when reading. The reason for it existing, the `join` call, comes after you needed to know why the list is there, which is a slight stumbling block for readability.

**Comment by tatumizer** [comment]
An accident indeed, and a happy one at that :smile:

For anyone familiar with the idiom, it's quite readable. (To me, it feels more readable than the alternatives.)

Performance-wise, the compiler can intrinsify it , thus avoiding an extra allocation, and/or implement other optimizations. There might not be a whole lot of overhead in proportion to other expenses in the above example anyway.
